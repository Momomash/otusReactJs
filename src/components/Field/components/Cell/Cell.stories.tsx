import * as React from 'react';
import { action } from '@storybook/addon-actions';
import { withKnobs, text, number } from '@storybook/addon-knobs';
import { Cell, CellStatus } from './Cell';

export default {
    title: 'Cell',
    decorators: [withKnobs],
};

export const Empty = () => [
    <Cell onClick={action('Cell Empty clicked')} status={CellStatus.Empty} key="jsx" />,
];
export const Living = () => [
    <Cell onClick={action('Cell Living clicked')} status={CellStatus.Living} key="jsx" />,
];
export const Young = () => [
    <Cell onClick={action('Cell Young clicked')} status={CellStatus.Young} key="jsx" />,
];