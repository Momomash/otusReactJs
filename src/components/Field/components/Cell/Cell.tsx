import * as React from 'react';

import styled from '@emotion/styled';

export enum CellStatus {
    Empty = 0,
    Living = 1,
    Young = 2,
}

const CellDiv = styled.div((props: { status: CellStatus; isAnimation: boolean }) => ({
    backgroundColor:
        props.status === CellStatus.Empty
            ? 'white'
            : props.status === CellStatus.Living
            ? '#ebded3'
            : '#584F4A',
    border: '1px solid #FCF1E4',
    width: '25px',
    height: '25px',
    transitionDuration: props.isAnimation ? '0.5s' : '0s',
}));

type Props = {
    status: CellStatus;
    onClick: () => void;
};
type State = {
    isAnimation: boolean;
};

export class Cell extends React.Component<Props, State> {
    state = { isAnimation: false };
    componentDidUpdate(prevProps: Props): void {
        const { status } = this.props;
        if (prevProps.status !== status) {
            if (status !== CellStatus.Empty) {
                this.setState({ isAnimation: true });
            } else {
                this.setState({ isAnimation: false });
            }
        }
    }
    render(): JSX.Element {
        return (
            <CellDiv
                isAnimation={this.state.isAnimation}
                status={this.props.status}
                onClick={(): void => this.props.onClick()}
            />
        );
    }
}
