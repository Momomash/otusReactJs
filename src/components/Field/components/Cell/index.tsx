export { CellStatus, Cell} from './Cell';
