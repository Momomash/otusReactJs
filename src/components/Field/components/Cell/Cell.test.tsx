import * as React from 'react';
import { mount } from 'enzyme';
import * as renderer from 'react-test-renderer';

import { Cell, CellStatus } from './Cell';

describe('Cell', () => {
    it('render Empty Cell', () => {
        const status = CellStatus.Empty;
        expect(
            renderer.create(<Cell onClick={jest.fn()} status={status} />).toJSON(),
        ).toMatchSnapshot();
    });
    it('render Living Cell', () => {
        const status = CellStatus.Living;
        expect(
            renderer.create(<Cell onClick={jest.fn()} status={status} />).toJSON(),
        ).toMatchSnapshot();
    });
    it('calls onClick callback with passed status', () => {
        const onClick = jest.fn();
        const status = CellStatus.Empty;
        const wrapper = mount(<Cell onClick={onClick} status={status} />);
        wrapper.simulate('click');
        expect(onClick).toHaveBeenCalledWith();
    });
});
