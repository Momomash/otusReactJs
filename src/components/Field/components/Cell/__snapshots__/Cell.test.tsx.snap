// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell render Empty Cell 1`] = `
.emotion-0 {
  background-color: white;
  border: 1px solid lightgray;
  width: 20px;
  height: 20px;
}

<div
  className="emotion-0"
  onClick={[Function]}
/>
`;
